### PR TITLE
feat: technical team type input is a radio button list

### PR DESCRIPTION
**Pull request by rdubigny**

### CODE REVIEW

**Comment by rdubigny** [comment]
[ETQ U, je vois le champ éditeur avec son nouveau design](https://trello.com/c/ghvp4Sd9/1292-etq-u-je-vois-le-champ-%C3%A9diteur-avec-son-nouveau-design)

File: frontend/src/components/templates/Form/index.js [hunk 1]
```diff
@@ -152,11 +174,22 @@ export const Form = ({
...
+            showOnlyFirstStep,
+            setShowOnlyFirstStep,
+            isShowNextStepButtonDisabled,
+            setIsShowNextStepButtonDisabled,
           }}
         >
           <HeadSection />
           <StepperSection />
-          {children}
+          {children
```
Comment (rdubigny) [inline frontend/src/components/templates/Form/index.js:185]: use destructuring if possible instead of slice

File: frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js [hunk 2]
```diff
@@ -44,33 +85,64 @@ export const TechnicalTeamCard = ({ editorList = [] }) => {
...
+            label="Qui s’occupera des aspects techniques et informatiques ?"
+            options={[{ id: '', label: selectedTypeLabel }]}
+            name="technical_team_type"
+            value=""
+            disabled={disabled}
+            onChange={() => null}
+            required
+          />
+        </>
       )}
```
Comment (JulieBoulangee) [inline frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js:132]: This code is rather long, I'd factorize this way : 
```
{technical_team_type ?
        !disabled &&
          <div
            style={{
              marginBottom: '0.5rem',
            }}
          >
            <Button onClick={resetType} icon="arrow-left" outline>
              retour
            </Button>
          </div>
        :
        <Alert>
          Pour lire les données vous aurez besoin d’un logiciel ou d’un
          service en ligne. <br />
          <Link href="/faq#champ-editeur" inline newTab>
            En savoir plus
          </Link>
        </Alert>
  }
  <RadioInput
    label="Qui s’occupera des aspects techniques et informatiques ?"
    options={technical_team_type ? typeOptions : [{ id: '', label: selectedTypeLabel }]}
    name="technical_team_type"
    value={technical_team_type || ""}
    disabled={disabled}
    onChange={technical_team_type ? onTypeChange : () => null}
    required
  />
```

**Review by JulieBoulangee — COMMENTED** [review]
Base de discussion pour entretien du 04/08/2022

File: frontend/src/components/molecules/ExpandableQuote/index.tsx [hunk 3]
```diff
@@ -1,22 +1,37 @@
...
+  const ref = useRef(null);
+
+  useEffect(() => {
+    if (scrollOnMount) {
+      // @ts-ignore
+      ref.current.scrollIntoView();
+    }
+  }, [ref, scrollOnMount]);
 
   useEffect(() => setExpanded(openOnMount), [openOnMount]);
```
Comment (hugues-m) [inline frontend/src/components/molecules/ExpandableQuote/index.tsx:30]: Not in this PR scope but this useEffect is not required. The goal is to have `expanded` initially set to the `openOnMount` prop. `useState(openOnMount)` is sufficient line 20

File: frontend/src/components/molecules/ExpandableQuote/index.tsx [hunk 4]
```diff
@@ -1,22 +1,37 @@
...
 
 export const ExpandableQuote: React.FC<Props> = ({
   title,
   large = false,
   openOnMount = false,
+  scrollOnMount = false,
   children,
 }) => {
   const [expanded, setExpanded] = useState(false);
+  const ref = useRef(null);
```
Comment (hugues-m) [inline frontend/src/components/molecules/ExpandableQuote/index.tsx:21]: `useRef<HTMLDivElement>(null)`  to type the ref according to its jsx usage

File: frontend/src/components/molecules/ExpandableQuote/index.tsx [hunk 5]
```diff
@@ -1,22 +1,37 @@
...
+
+  useEffect(() => {
+    if (scrollOnMount) {
+      // @ts-ignore
+      ref.current.scrollIntoView();
```
Comment (hugues-m) [inline frontend/src/components/molecules/ExpandableQuote/index.tsx:26]: No need for  ts-ignore if the type is set l.21. 
`ref.current?.scrollIntoView()` to be safe

File: frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js [hunk 6]
```diff
@@ -1,22 +1,63 @@
...
   const {
     disabled,
     onChange,
+    isUserEnrollmentLoading,
     enrollment: { technical_team_type, technical_team_value },
   } = useContext(FormContext);
 
+  const { setReadyForNextSteps } = useContext(HideSectionsContext);
+
+  const areInputValid = useMemo(() => {
```
Comment (hugues-m) [inline frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js:28]: Premature optimisations with useMemo make the code harder to read and maintain, is his not more performant for light computation due to useMemo overhead.
 See the deep dive section at https://beta.reactjs.org/learn/you-might-not-need-an-effect#caching-expensive-calculations
 
Just use boolean checks : 
```ts
const areInputValid =
    technical_team_type === 'internal_team' || !!technical_team_value;```

File: frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js [hunk 7]
```diff
@@ -1,22 +1,63 @@
...
+    if (technical_team_type === 'internal_team') {
+      return true;
+    }
+    if (technical_team_value) {
+      return true;
+    }
+    return false;
+  }, [technical_team_type, technical_team_value]);
+
+  useEffect(() => {
```
Comment (hugues-m) [inline frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js:41]: This logic is a state update depending on a form event. It should be handled in a callback called by the Form on form state change.
I see that it is not available yet in the form hook used here si this is not in the scope of this PR.

File: frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js [hunk 8]
```diff
@@ -1,22 +1,63 @@
...
+    !isUserEnrollmentLoading && setReadyForNextSteps(areInputValid);
+  }, [isUserEnrollmentLoading, setReadyForNextSteps, areInputValid]);
+
   const editorOptions = useMemo(
     () => editorList.map(({ siret, name }) => ({ id: siret, label: name })),
     [editorList]
   );
 
+  const selectedTypeLabel = useMemo(
```
Comment (hugues-m) [inline frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js:50]: useMemo overhead not worth it 
```ts
  const selectedTypeLabel = typeOptions.find(
    ({ id }) => id === technical_team_type
  )?.label;
  ```

File: frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js [hunk 9]
```diff
@@ -44,33 +85,64 @@ export const TechnicalTeamCard = ({ editorList = [] }) => {
...
-          label="Quel est votre éditeur de logiciel ?"
-          options={editorOptions}
-          name="technical_team_value"
-          value={technicalTeamValueLabel}
-          disabled={disabled}
-          onChange={onValueChange}
-          required
-        />
+      <h3>Qui implémentera l’API ?</h3>
+      {!isUserEnrollmentLoading && !technical_team_type && (
```
Comment (hugues-m) [inline frontend/src/components/organisms/form-sections/OrganisationSection/TechnicalTeamCard.js:89]: There is mutually exclusive contents depending on isUserEnrollmentLoading and technical_team_type
The code do not reflect intents and the conditions are a bit hard to follow.
Also in general keep this kind of logic out of the jsx
I provide here an example using smaller logic blocks but cleaner implementation with other component(s) may be better.

```tsx

  const technicalTeamTypeInput = !!technical_team_type ? (
    <>
      {disabled ? null : (
        <div
          style={{
            marginBottom: '0.5rem',
          }}
        >
          <Button onClick={resetType} icon="arrow-left" outline>
            retour
          </Button>
        </div>
      )}
      <RadioInput
        label="Qui s’occupera des aspects techniques et informatiques ?"
        options={[{ id: '', label: selectedTypeLabel }]}
        name="technical_team_type"
        value=""
        disabled={disabled}
        onChange={() => null}
        required
      />
    </>
  ) : (
    <>
      <Alert>
        Pour lire les données vous aurez besoin d’un logiciel ou d’un service en
        ligne. <br />
        <Link href="/faq#champ-editeur" inline newTab>
          En savoir plus
        </Link>
      </Alert>
      <RadioInput
        label="Qui s’occupera des aspects techniques et informatiques ?"
        options={typeOptions}
        name="technical_team_type"
        value={technical_team_type}
        disabled={disabled}
        onChange={onTypeChange}
        required
      />
    </>
  );

  // Display relevant technicalTeamValue input depending on teamType
  const technicalTeamValueInput =
    technical_team_type === 'software_company' ? (
      <TextInputWithSuggestions
        label="Quel est votre éditeur de logiciel ?"
        options={editorOptions}
        name="technical_team_value"
        value={technicalTeamValueLabel}
        disabled={disabled}
        onChange={onValueChange}
        required
      />
    ) : technical_team_type === 'other' ? (
      <TextInput
        label="Précisez ?"
        name="technical_team_value"
        value={technical_team_value}
        disabled={disabled}
        onChange={onChange}
        required
      />
    ) : null;

  const cardContent = isUserEnrollmentLoading ? null : (
    <>
      {technicalTeamTypeInput}
      {technicalTeamValueInput}
    </>
  );

  return (
    <Card>
      <h3>Qui implémentera l’API ?</h3>
      {cardContent}
    </Card>
  );
  ```

File: frontend/src/components/templates/Faq.tsx [hunk 10]
```diff
@@ -157,6 +158,7 @@ const FAQ: React.FC<{}> = () => (
           <ExpandableQuote
             title={question.title}
             openOnMount={window.location.hash === `#${question.anchor}`}
+            scrollOnMount={window.location.hash === `#${question.anchor}`}
```
Comment (hugues-m) [inline frontend/src/components/templates/Faq.tsx:161]: Keep logic outside of jsx 
```tsx
{questions.map((question) => {
        const isNavigatingToQuestion =
          window.location.hash === `#${question.anchor}`;

        return (
          <div
            key={question.anchor}
            id={question.anchor}
            style={{ marginBottom: '1.5rem' }}
          >
            <ExpandableQuote
              title={question.title}
              openOnMount={isNavigatingToQuestion}
              scrollOnMount={isNavigatingToQuestion}
            >
              {question.body}
            </ExpandableQuote>
          </div>
        );
      })}
      ```

File: frontend/src/components/templates/Form/HideSectionsContainer.tsx [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  SubmissionPanel,
+  children,
+}) => {
+  const [showOnlyFirstStep, setShowOnlyFirstStep] = useState(false);
+  const [isShowNextStepButtonDisabled, setIsShowNextStepButtonDisabled] =
+    useState(false);
+  const [areNextStepsLoading, setAreNextStepLoading] = useState(false);
+  const [readyForNextSteps, setReadyForNextSteps] = useState(true);
+
+  useEffect(() => {
```
Comment (hugues-m) [inline frontend/src/components/templates/Form/HideSectionsContainer.tsx:33]: This kind of useEffect is discouraged
See the Recap for useEffect doc here : https://beta.reactjs.org/learn/synchronizing-with-effects#recap
 - read the whole page if something is not clear
 - please do the challenges in the Recap section

The 2 useEffect in this component should be removed. The second one also has the same dependencies so will be triggered immediately following the first one. That could cause unexpected side effects and is hard to follow.

File: frontend/src/components/templates/Form/HideSectionsContainer.tsx [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+  const showNextStep = () => {
+    setIsShowNextStepButtonDisabled(true);
+    setAreNextStepLoading(true);
+    setTimeout(() => {
+      setAreNextStepLoading(false);
+      setShowOnlyFirstStep(false);
+    }, 1800);
+  };
+
+  useEffect(() => {
```
Comment (hugues-m) [inline frontend/src/components/templates/Form/HideSectionsContainer.tsx:48]: Same as above, this useEffect should be removed.

You have 2 handlers for setting `showOnlyFirstStep` and `readyForNextSteps` that have side effect. It should be explicit in the code.

Do not use the `useState` setters if there are side effect on other states or logic to do. Instead use your own handlers.

Here is an example implementation that should replace the call for the useState setters in this component and jsx: 
```ts

  const handleOnlyShowFirstStep = (show: boolean) => {
    if (show) {
      setIsOnlyFirstStepShown(true);
      setIsShowNextStepButtonDisabled(!isReadyForNextSteps);
      return;
    }

    setIsOnlyFirstStepShown(!isReadyForNextSteps);
  };
  ```
```ts
  const handleReadyForNextStep = (ready: boolean) => {
    if (isOnlyFirstStepShown) {
      setIsShowNextStepButtonDisabled(!ready);
      return;
    }

    onlyShowFirstStep(!ready);
  };
  ```

File: frontend/src/components/templates/Form/HideSectionsContainer.tsx [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  const [areNextStepsLoading, setAreNextStepLoading] = useState(false);
+  const [readyForNextSteps, setReadyForNextSteps] = useState(true);
+
+  useEffect(() => {
+    if (showOnlyFirstStep) {
+      setIsShowNextStepButtonDisabled(true);
+    }
+  }, [setIsShowNextStepButtonDisabled, showOnlyFirstStep]);
+
+  const showNextStep = () => {
```
Comment (hugues-m) [inline frontend/src/components/templates/Form/HideSectionsContainer.tsx:39]: 👍 this is the right way to do it, an handler that have the logic and mutate multiple state hooks

File: frontend/src/components/templates/Form/index.js [hunk 14]
```diff
@@ -158,7 +159,17 @@ export const Form = ({
         >
           <HeadSection />
           <StepperSection />
-          {children}
+          <HideSectionsContainer
+            SubmissionPanel={() => (
```
Comment (hugues-m) [inline frontend/src/components/templates/Form/index.js:163]: This code is equivalent to creating a functional component "on the fly" and passing it as a prop, delegating the responsibility of instantiation to `HideSectionsContainer`.
This will not work as this callback will be different on every render and will create a new component.

Remember that the code between jsx tags (children), is in fact just another prop.
You can use your own props to pass "other children" or "slots".

Also props should be in camelCase

Replace this code with this implementation and change HideSectionsContainer accordingly

```tsx
<HideSectionsContainer
            submissionPanel={
              <SubmissionPanel
                enrollment={enrollment}
                handlePostEvent={handlePostEvent}
                updateEnrollment={dispatchSetEnrollment}
              />
            }
          >
 ```

File: frontend/src/components/templates/Form/HideSectionsContainer.tsx [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+export type HideSectionsContextType = {
+  setReadyForNextSteps: (v: boolean) => void;
+};
+
+export const HideSectionsContext = createContext<HideSectionsContextType>({
+  setReadyForNextSteps: () => null,
+});
+
+type Props = {
+  SubmissionPanel: FunctionComponent;
```
Comment (hugues-m) [inline frontend/src/components/templates/Form/HideSectionsContainer.tsx:20]: (See comment on Form component)
```ts
{ submissionPanel: ReactNode}
```

File: frontend/src/components/templates/Form/HideSectionsContainer.tsx [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+            Suivant
+          </Button>
+        </div>
+      )}
+
+      {showOnlyFirstStep && areNextStepsLoading && (
+        <Loader message="Nous pré-remplissons le formulaire en fonction de vos réponses." />
+      )}
+
+      {!showOnlyFirstStep && <SubmissionPanel />}
```
Comment (hugues-m) [inline frontend/src/components/templates/Form/HideSectionsContainer.tsx:105]: (See comment on Form component)
```ts
{!showOnlyFirstStep && submissionPanel}
```

**Review by hugues-m — COMMENTED** [review]
This PR seems functional, except for a bug in HideSectionContainer/SubmissionPanel integration.
There is opportunity to review our React guidelines from this PR, please reach out to go over this together.
